nanopub/publication.py

```diff
@@ -121,83 +100,119 @@ def from_assertion(cls, assertion_rdf: rdflib.Graph,
                     main_graph.bind(prefix, namespace)
-                cls._replace_blank_nodes(dummy_namespace=this_np, rdf=user_rdf)
+                cls._replace_blank_nodes(rdf=user_rdf)
         assertion += assertion_rdf
         if provenance_rdf is not None:
             provenance += provenance_rdf
         if pubinfo_rdf is not None:
             pubinfo += pubinfo_rdf
 
         creationtime = rdflib.Literal(datetime.now(), datatype=XSD.dateTime)
-        provenance.add((this_np.assertion, namespaces.PROV.generatedAtTime, creationtime))
-
-        pubinfo.add((this_np[''], namespaces.PROV.generatedAtTime, creationtime))
+        provenance.add((DUMMY_NAMESPACE.assertion, namespaces.PROV.generatedAtTime, creationtime))
+        pubinfo.add((DUMMY_NAMESPACE[''], namespaces.PROV.generatedAtTime, creationtime))
 
         if assertion_attributed_to:
-            assertion_attributed_to = rdflib.URIRef(assertion_attributed_to)
-            provenance.add((this_np.assertion,
-                            namespaces.PROV.wasAttributedTo,
-                            assertion_attributed_to))
+            cls._handle_assertion_attributed_to(assertion_attributed_to, provenance)
 
         if derived_from:
-            uris = []
-            if isinstance(derived_from, list):
-                list_of_URIs = derived_from
-            else:
-                list_of_URIs = [derived_from]
+            cls._handle_derived_from(derived_from, pubinfo)
 
-            for derived_from_uri in list_of_URIs:
-                # Convert uri to an rdflib term first (if necessary)
-                derived_from_uri = rdflib.URIRef(derived_from_uri)
-
-                provenance.add((this_np.assertion,
-                                namespaces.PROV.wasDerivedFrom,
-                                derived_from_uri))
+        if introduces_concept:
+            cls._handle_introduces_concept(introduces_concept, pubinfo)
 
         # Always attribute the nanopublication (not the assertion) to the ORCID iD in user profile
-        pubinfo.add((this_np[''],
-                      namespaces.PROV.wasAttributedTo,
-                      rdflib.URIRef(profile.get_orcid_id())))
+        pubinfo.add((DUMMY_NAMESPACE[''],
+                     namespaces.PROV.wasAttributedTo,
+                     rdflib.URIRef(profile.get_orcid_id())))
 
-        if introduces_concept:
-            # Convert introduces_concept URI to an rdflib term first (if necessary)
-            if isinstance(introduces_concept, rdflib.term.BNode):
-                introduces_concept = this_np[str(introduces_concept)]
-            else:
-                introduces_concept = rdflib.URIRef(introduces_concept)
+        return cls(rdf=main_graph)
 
-            pubinfo.add((this_np[''],
-                          namespaces.NPX.introduces,
-                          introduces_concept))
+    @staticmethod
+    def _handle_assertion_attributed_to(assertion_attributed_to, provenance):
```

Comment:
These utility methods definitely make ```from_assertion()``` easier to parse by eye, so we should keep them. My only issue is that now all these 'private' methods are cluttering the ```Publication``` class. Then again, I think moving them into their own module is more confusing, since the code is more understandable if these helpers are close to the (only) method they act in. Perhaps give them a docstring saying 'handler for from_assertion()' to make it clear which method they belong to.

Reply:
It depends on how you view it. This code was already in there, only now it is clearer what it does. Indeed it makes the class more cluttered, but the `from_assertion` method less.  But I added the 'handler for from_assertion' docstrings that is a nice suggestion!